packages/data/addon/services/elide-metadata.ts

```diff
@@ -0,0 +1,273 @@
+   * @param type
+   * @param id
+   * @param namespace - optional
+   * @returns {MetadataModel} metadata model object
+   */
+  getById(type: MetadataType, id: string, namespace?: string) {
+    assert('Type must be a valid navi-data model type', this.isValidType(type));
+    let source = namespace || getDefaultDataSourceName();
+    assert('Metadata must be loaded before the operation can be performed', this.loadedDataSources.includes(source));
+
+    return this._keg.getById(`metadata/${type}`, id, source);
+  }
+
+  /**
+   * @method fetchById
+   * Fetch metadata based on type and id
+   *
+   * @param type
+   * @param id
+   * @param namespace
+   * @returns fetched record
+   */
+  fetchById(/*type: MetadataQueryType, id: string, namespace: string*/) {
+    //TODO: Implement fetchById
+    return undefined;
+  }
+
+  /**
+   * @method findById
+   * gets Metadata or fetches it if necessary
+   *
+   * @param type
+   * @param id
+   * @param namespace
+   * @returns {Promise<MetadataModel>}
+   */
+  findById(type: MetadataType, id: string, namespace?: string) {
+    //Get entity if already present in the keg
+    let dataSourceName = namespace || getDefaultDataSourceName();
+
+    // TODO: fallback to fetch when fetchById is supported
+    return Promise.resolve(this.getById(type, id, dataSourceName));
+  }
+
+  /**
+   * Convenience method to get a meta data field
+   * @param type
+   * @param id
+   * @param field
+   * @param defaultIfNone - (optional) default if meta data or field isn't found
+   * @param namespace
+   * @returns field if found otherwise defaultIfNone
+   */
+  getMetaField(
+    type: MetadataType,
+    id: string,
+    field: string,
+    defaultIfNone: string | null = null,
+    namespace: string | null = null
+  ) {
+    const dataSourceName = namespace || getDefaultDataSourceName();
+    const meta = this.getById(type, id, dataSourceName);
+    return meta?.[field] || defaultIfNone;
+  }
+
+  /**
+   * Convenience method to get namespace of a table
+   * @param {String} table
+   * @returns {string} - namespace
+   */
+  getTableNamespace(tableId: string) {
+    const items = this._keg.getBy('metadata/table', (recordTable: TableMetadata) => recordTable.id === tableId);
+
+    return items.length ? items[0].source : getDefaultDataSourceName();
+  }
+
+  /**
+   * Runtime typecheck that typescript can understand
+   * @param metadata - Payload
+   * @returns true if metadata is of type NormalizedMetadata
+   */
+  isMetadataNormalized(metadata: NormalizedMetadata | TablePayload | undefined): metadata is NormalizedMetadata {
+    return Object.keys(metadata || {}).every(type =>
+      VALID_TYPES.includes(singularize(dasherize(type)) as MetadataType)
+    );
+  }
+
+  /**
+   * Runtime typecheck that typescript can understand
+   * @param type - type name
+   * @returns true if type is MetadataType
+   */
+  isValidType(type: string): type is MetadataType {
+    return VALID_TYPES.includes(type as MetadataType);
```

Comment:
https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates

Typescript complains if we don't cast it.
`Argument of type 'string' is not assignable to parameter of type '"table" | "metric" | "dimension" | "time-dimension"'`

I don't think casting it is an issue because it will return true or false correctly regardless.

Reply:
Yea I'd just prefer to avoid casting when possible. I think I'd prefer to cast `VALID_TYPES as string[]` instead of doing a possibly invalid cast